src/main/java/io/projectriff/invoker/JavaFunctionInvokerServer.java

```diff
@@ -66,33 +63,24 @@ public JavaFunctionInvokerServer(Function<Flux<?>, Flux<?>> function, Gson mappe
 
 	private Function<Flux<Message<?>>, Flux<Message<?>>> enhance(
 			Function<Flux<Message<?>>, Flux<Message<?>>> function) {
-		return messages -> function.apply(messages.flatMap(m -> push(m)))
-				.flatMap(m -> pop(m)).subscriberContext(setup());
-	}
-
-	private Function<Context, Context> setup() {
-		return ctx -> ctx.put(CONTEXT, new HashMap<String, Object>());
+		AtomicReference<Map<String, Object>> headers = new AtomicReference<>();
+		return messages -> function.apply(messages.map(m -> push(headers, m)))
+				.map(m -> pop(headers, m));
 	}
 
-	private Mono<Message<?>> pop(Message<?> message) {
-		return Mono.subscriberContext().map(ctx -> {
-			Map<String, Object> map = ctx.get(CONTEXT);
-			if (map.isEmpty()) {
-				return message;
-			}
-			Message<?> result = MessageBuilder.fromMessage(message)
-					.copyHeadersIfAbsent(map).build();
-			map.clear();
-			return result;
-		});
+	private Message<?> pop(AtomicReference<Map<String, Object>> headers,
+			Message<?> message) {
+		Map<String, Object> map = headers.getAndSet(Collections.emptyMap());
+		if (map.isEmpty()) {
+			return message;
+		}
+		return MessageBuilder.fromMessage(message).copyHeadersIfAbsent(map).build();
 	}
 
-	private Mono<Message<?>> push(Message<?> value) {
-		return Mono.subscriberContext().map(ctx -> {
-			Map<String, Object> map = ctx.get(CONTEXT);
-			map.putAll(value.getHeaders());
-			return value;
-		});
+	private Message<?> push(AtomicReference<Map<String, Object>> headers,
```

Comment:
I find the method names `push` and `pop` a bit confusing as there is no stack. Please would you consider renaming them. Suggestion: `storeHeaders` and `retrieveHeaders`.

(Also, even though it's outside the scope of this PR, would `enhance` be better named `preserveHeaders`?)

Reply:
Done (all as suggested) in commit to follow. I didn't call it "preserveHeaders" originally because it was going to be given more responsibility, but then didn't need it, apparently.